platforms/raspi/raspi_adaptor.go

```diff
@@ -100,18 +116,31 @@ func (r *Adaptor) DigitalPin(pin string, dir string) (sysfsPin sysfs.DigitalPinn
 		return
 	}
 
-	if r.digitalPins[i] == nil {
-		r.digitalPins[i] = sysfs.NewDigitalPin(i)
-		if err = r.digitalPins[i].Export(); err != nil {
-			return
-		}
+	currentPin, err := r.getExportedDigitalPin(i, dir)
+
+	if err != nil {
+		return
 	}
 
-	if err = r.digitalPins[i].Direction(dir); err != nil {
+	if err = currentPin.Direction(dir); err != nil {
 		return
 	}
 
-	return r.digitalPins[i], nil
+	return currentPin, nil
+}
+
+func (r *Adaptor) getExportedDigitalPin(translatedPin int, dir string) (sysfsPin sysfs.DigitalPinner, err error) {
```

Comment:
in `PWMPin()` I preferred an explicit locking strategy (lines 199 and 200) without creating a new func because in this case we're just creating a new struct and returning it immediately.